libs/libc/uio/lib_readv.c

```diff
@@ -68,56 +70,63 @@
 
 ssize_t readv(int fildes, FAR const struct iovec *iov, int iovcnt)
 {
-  ssize_t ntotal;
-  ssize_t nread;
-  size_t remaining;
   FAR uint8_t *buffer;
+  ssize_t nread;
+  size_t tocopy;
+  size_t total_size;
+  size_t offset;
   int i;
 
-  /* Process each entry in the struct iovec array */
-
-  for (i = 0, ntotal = 0; i < iovcnt; i++)
+  if (iovcnt == 0)
     {
-      /* Ignore zero-length reads */
-
-      if (iov[i].iov_len > 0)
-        {
-          buffer    = iov[i].iov_base;
-          remaining = iov[i].iov_len;
-
-          /* Read repeatedly as necessary to fill buffer */
-
-          do
-            {
-              /* NOTE:  read() is a cancellation point */
-
-              nread = read(fildes, buffer, remaining);
```

Comment:
why we can't read the data piece by piece?

Reply:
because, in general, you should not keep reading after a short read.

Reply:
yes, we can break out by returning the length we really got so far.

Reply:
BTW, I don't get the problem from your commit message. could you explain more?

Reply:
consider that you readv() on a tty, using a huge buffer, say, 1024 bytes.
if you got some data, say, 1 byte, readv() should return the 1 byte without waiting for the following 1023 bytes.

Reply:
> yes, we can break out by returning the length we really got so far.

i don't think you can do that reliably because data boundary can happen to match iov boundary.

Reply:
the loop can be changed to:
```
ssize_t readv(int fildes, FAR const struct iovec *iov, int iovcnt)
{
  ssize_t ntotal;
  ssize_t nread;
  int i;

  /* Process each entry in the struct iovec array */

  for (i = 0, ntotal = 0; i < iovcnt; i++)
    {
      /* Ignore zero-length reads */

      if (iov[i].iov_len > 0)
        {
          nread = read(iov[i].iov_base, iov[i].iov_len);
          if (nread <= 0)
              {
                break;
              }

          buffer  += nread;
          ntotal  += nread;

          if (nread != iov[i].iov_len)
              {
                break;
              }
        }
    }

  return ntotal ? ntotal : nread;
}
```

Reply:
> the loop can be changed to:

it seems broken to me.

consider:
* iovcnt == 2
* iov[0]->iov_len == 4
* iov[1]->iov_len == 4
* 4 byte data is available to read from the fd

Reply:
changing to nonblock mode or calling select/poll with 0 timeout could fix this case.

Reply:
no.

* these tricks are not thread-safe.
* we should preserve packet boundary for some kind of fds like datagram socket.

Reply:
the only sane way w/o malloc i can think of is to make the kernel implement readv natively and implement read on the top of it. but it would be a far more invasive change than this PR. (probably we need to modify every drivers) thus i submitted this PR for now.

Reply:
if you are so unhappy with malloc, how about doing stat() and fallback to the old implementation for regular files? i guess it covers ~90% of cases. i personally prefer to avoid such "optimization" for the sake of simplicity though.

Reply:
another compatible implementation is always just read the first element.

Reply:
> another compatible implementation is always just read the first element.

i don't understand what you mean. can you explain?

Reply:
like this:
```
ssize_t readv(int fildes, FAR const struct iovec *iov, int iovcnt)
{
  return iovcnt ? read(fildes, iov->iov_base, iov->iov_len) : 0;
}
```

Reply:
> like this:
> 
> ```
> ssize_t readv(int fildes, FAR const struct iovec *iov, int iovcnt)
> {
>   return iovcnt ? read(fildes, iov->iov_base, iov->iov_len) : 0;
> }
> ```

what is it compatible with?
to me it seems broken even for regular files.

Reply:
compatible with the spec, the spec never requires we return all available data:
https://pubs.opengroup.org/onlinepubs/009695399/functions/readv.html
so, the implementation avoids the allocation and conform to the spec.

Reply:
i'm not sure if such an interpretation of the spec makes much sense.
as you know, regular files are usually not interruptible on unix-like operating systems and many applications are relying on it.

Reply:
@yamt @xiaoxiang781216 how Linux, QNX, MacOS, etc implement it? Although the specification doesn't requires some assumptions, it is important to follow what other OSes are doing, to avoid  breaking applications when porting it to NuttX

Reply:
I agree that there is an issue with `readv` behaviour, but adding `malloc` and trying to avoid memory leak with nested cancellation point is also conceptually wrong. I do not have a better solution right not and just noticed and `writev` seems also to suffer from the same problem.

Reply:
Sounds like `struct file_operations` read and write methods should actually operate with vectors (the same as `file_read` and `file_write`), but such a change basically would require rewriting thousands lines of the code...

Reply:
> @yamt @xiaoxiang781216 how does Linux, QNX, MacOS, etc implement it? Although the specification doesn't require some assumptions, it is important to follow what other OSes are doing, to avoid breaking applications when porting it to NuttX

Linux improve the driver model by changing argument from void * to iovec*.

Reply:
> no.
> 
> * these tricks are not thread-safe.

> * we should preserve packet boundary for some kind of fds like datagram socket.

socket already improve this case by: https://github.com/apache/nuttx/pull/2959. But file_operation_s doesn't have the similar patch, so it's hard to modify readv to utilize this new capability from socket layer.

Reply:
> but adding `malloc` and trying to avoid memory leak with nested cancellation point is also conceptually wrong.

it's just inefficient.
it's far better than the original implementation, which is broken in a user-visible way.

> I do not have a better solution right not and just noticed and `writev` seems also to suffer from the same problem.

right. writev/preadv/pwritev have the same problem.

Reply:
if someone plans to do the ideal solution (ie. make ~everything iovec-based) anytime soon, it's great.
otherwise, i'd suggest to merge this PR for now.